bundles/io/org.eclipse.smarthome.io.transport.mdns/src/main/java/org/eclipse/smarthome/io/transport/mdns/internal/MDNSClientImpl.java

```diff
@@ -24,25 +31,60 @@
 public class MDNSClientImpl implements MDNSClient {
     private final Logger logger = LoggerFactory.getLogger(MDNSClientImpl.class);
 
-    private JmDNS jmdns;
+    private Set<JmDNS> jmdnsInstances = new CopyOnWriteArraySet<>();
+
+    private static Set<InetAddress> getAllInetAddresses() {
+        final Set<InetAddress> addresses = new HashSet<>();
+        Enumeration<NetworkInterface> itInterfaces;
+        try {
+            itInterfaces = NetworkInterface.getNetworkInterfaces();
+        } catch (final SocketException e) {
+            return addresses;
+        }
+        while (itInterfaces.hasMoreElements()) {
+            final NetworkInterface iface = itInterfaces.nextElement();
+            try {
+                if (!iface.isUp() || iface.isLoopback()) {
+                    continue;
+                }
+            } catch (final SocketException ex) {
+                continue;
+            }
+            final Enumeration<InetAddress> itAddresses = iface.getInetAddresses();
+            while (itAddresses.hasMoreElements()) {
+                final InetAddress address = itAddresses.nextElement();
+                if (address.isLoopbackAddress() || address.isLinkLocalAddress()) {
+                    continue;
+                }
+                addresses.add(address);
+            }
+        }
+        return addresses;
+    }
 
     @Override
-    public JmDNS getClient() {
-        return jmdns;
+    public Set<JmDNS> getClientInstances() {
```

Comment:
Did you check whether this API change breaks any existing bindings in openHAB 2?
If it does NOT break any, it means that getClient() is only used by the MDNSDiscoveryService implementation. If this is indeed the case, I wonder if there is at all a need to expose the full set of JmDNS instances. Maybe we could simply return some simple wrapper that mimicks a JmDNS instance, but forwards the calls (e.g. to addServiceListener) to all internal instances. This would make the code for the caller much simpler, as there wouldn't be a need for loops anymore.